targets/TARGET_ARM_FM/TARGET_FVP_MPS2/spi_api.c

```diff
@@ -61,278 +60,262 @@ static const PinMap PinMap_SPI_SSEL[] = {
+                      SSP_CR0_DSS_8;      /* 8 bit symbol size                */
+    obj->spi->CPSR  = SSP_CPSR_DFLT;
+    obj->spi->DMACR = 0;                  /* Disable FIFO DMA                 */
+    obj->spi->IMSC  = 0;                  /* Mask all FIFO/IRQ interrupts     */
+    obj->spi->ICR   = SSP_ICR_RORIC_Msk |
+                      SSP_ICR_RTIC_Msk;   /* Clear both interrupts            */
+
+    if (is_slave) {
+        obj->spi->CR1 |= SSP_CR1_SSE_Msk; /* Device configured as slave       */
+    } else {
+        obj->spi->CR1 &= ~SSP_CR1_SSE_Msk;/* Device configured as master      */
+    }
 
-int spi_slave_read(spi_t *obj)
-{
-    return obj->spi->DR;
-}
+    ssp_enable(obj);
 
-void spi_slave_write(spi_t *obj, int value)
-{
-    while (ssp_writeable(obj) == 0) ;
-    obj->spi->DR = value;
+    // pin out the spi pins
+    pinmap_pinout(mosi, PinMap_SPI_MOSI);
+    pinmap_pinout(miso, PinMap_SPI_MISO);
+    pinmap_pinout(sclk, PinMap_SPI_SCLK);
+    if (ssel != NC) {
+        pinmap_pinout(ssel, PinMap_SPI_SSEL);
+    }
 }
 
-int spi_busy(spi_t *obj)
+uint32_t spi_transfer(spi_t *obj, const void *tx, uint32_t tx_len, void *rx, uint32_t rx_len, const void *fill_symbol)
 {
-    return ssp_busy(obj);
+    uint32_t total = (tx_len > rx_len) ? tx_len : rx_len;
+
+    for (uint32_t i = 0; i < total; i++) {
+        uint16_t out, in;
+        switch (symbol_byte_size(obj)) {
+            case 1: // symbol size 8 bit or less
+                if (tx) {
+                    out = (i < tx_len) ? ((uint8_t *)tx)[i] : *(uint8_t *)fill_symbol;
+                } else {
+                    out = *(uint8_t *)fill_symbol;
+                }
+                if (_spi_bit_order == SPI_BIT_ORDERING_LSB_FIRST) {
+                    out = reverse_byte((uint8_t)out);
+                }
+                ssp_write(obj, out);
+                in = (uint8_t)ssp_read(obj);
+                if (i < rx_len) {
+                    if (_spi_bit_order == SPI_BIT_ORDERING_LSB_FIRST) {
+                        in = reverse_byte((uint8_t)in);
+                    }
+                    ((uint8_t *)rx)[i] = (uint8_t)in;
+                }
+                break;
+            case 2: // symbol size more than 8 bit
+                if (tx) {
+                    out = (i < tx_len) ? ((uint16_t *)tx)[i] : *(uint16_t *)fill_symbol;
+                } else {
+                    out = *(uint16_t *)fill_symbol;
+                }
+                if (_spi_bit_order == SPI_BIT_ORDERING_LSB_FIRST) {
+                    out = (reverse_byte(out & 0xFF)) << 8U | reverse_byte(out >> 8U);
+                }
+                ssp_write(obj, out);
+                in = ssp_read(obj);
+                if (i < rx_len) {
+                    if (_spi_bit_order == SPI_BIT_ORDERING_LSB_FIRST) {
+                        in = (reverse_byte(in & 0xFF)) << 8U | reverse_byte(in >> 8U);
+                    }
+                    ((uint16_t *)rx)[i] = in;
```

Comment:
It looks like NULL check is missing before data is stored into `rx`.
I suggest to modify line:
`if (i < rx_len) {`
to:
`if (rx && i < rx_len) {`